nixos/modules/security/systemd-confinement.nix

```diff
@@ -98,8 +105,11 @@ in {
         wantsAPIVFS = lib.mkDefault (config.confinement.mode == "full-apivfs");
       in lib.mkIf config.confinement.enable {
         serviceConfig = {
-          RootDirectory = "/var/empty";
-          TemporaryFileSystem = "/";
```

Comment:
I've refactored the NixOS test and while working on parametrising the subtests over root/static-user/dynamic-user I found that when using a static user, the root directory is now writable. This is obviously not what we want and setting `TemporaryFileSystem = "/";` here avoids this for both static and dynamic users.

So before I simply re-introduce this option: Is there a particular reason why you removed this?

Reply:
I just pushed the parametrised subtests so you can reproduce the issue.

Reply:
Oh, good catch! [I removed]((https://discourse.nixos.org/t/hardening-systemd-services/17147/14)) `TemporaryFileSystem = [ "/" ]` because [`ProtectSystem=` invalidates `TemporaryFileSystem=/`](https://github.com/systemd/systemd/issues/18999).
But then I did not notice that `/` was no longer necessarily owned by `root` but by the service's user, which is not a problem when `ProtectSystem=strict` (which is implied by `DynamicUser=true`), as `/` is then mounted read-only, but becomes a problem when only using `User=` (as it does not imply `ProtectSystem=strict`).

I can't explore solutions to that problem yet, but here are the details:

Using a `RuntimeDirectory=` as `RootDirectory=` has the side effect of using the service's user as the owner of `/`:
```console
$ sudo systemd-run -pUser=999 -P -pPrivateMounts=1 -pRuntimeDirectory=systemd-confinement/test -pRootDirectory=/run/systemd-confinement/test -pUMask=066 -pBindReadOnlyPaths=/nix/store -- $(realpath /bin/sh) -xc "PATH=$(dirname $(realpath -e $(which ls))); ls -ld /; mkdir -v /test"
Running as unit: run-u3733.service
+ PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin
+ ls -ld /
drwxr-xr-x 11 999 999 220 Apr 28 13:23 /
+ mkdir -v /test
mkdir: created directory '/test'
```

Whereas using `TemporaryFileSystem = [ "/" ]` has the side effect of making `root` the owner of `/`:
```console
$ sudo systemd-run -P -pUser=999 -pTemporaryFileSystem=/ -pPrivateMounts=1 -pRuntimeDirectory=systemd-confinement/test -pRootDirectory=/run/systemd-confinement/test -pUMask=066 -pBindReadOnlyPaths=/nix/store -- $(realpath /bin/sh) -xc "PATH=$(dirname $(realpath -e $(which ls))); ls -ld /; mkdir -v /test"
Running as unit: run-u3740.service
+ PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin
+ ls -ld /
drwxr-xr-x 7 0 0 140 Apr 28 13:28 /
+ mkdir -v /test
mkdir: cannot create directory '/test': Permission denied
```

The problem is not detected when using `DynamicUser=true` because it implies `ProtectSystem=strict`, making `/` read-only:
```console
$ sudo systemd-run -P -pUser=999 -pProtectSystem=strict -pPrivateMounts=1 -pRuntimeDirectory=systemd-confinement/test -pRootDirectory=/run/systemd-confinement/test -pUMask=066 -pBindReadOnlyPaths=/nix/store -- $(realpath /bin/sh) -xc "PATH=$(dirname $(realpath -e $(which ls))); ls -ld /; mkdir -v /test"
Running as unit: run-u3742.service
+ PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin
+ ls -ld /
drwxr-xr-x 11 999 999 220 Apr 28 13:29 /
+ mkdir -v /test
mkdir: cannot create directory '/test': Read-only file system
```

Reply:
> Oh, good catch! I removed `TemporaryFileSystem = [ "/" ]` because [`ProtectSystem=` invalidates `TemporaryFileSystem=/`](https://github.com/systemd/systemd/issues/18999).

So in theory we should just be able to leave `TemporaryFileSystem = "/";` and it should work for both cases since it will not change the behaviour for the `ProtectSystem`/`DynamicUser` case, do I understand this correctly?

Reply:
@aszlig, alas, though `TemporaryFileSystem=/` is (currently) compatible to some extent with `ProtectSystem=strict` (implied by `DynamicUser=true`), it is however incompatible with both `ProtectSystem=full` or `ProtectSystem=true`:

```console
$ sudo systemd-run -P -pTemporaryFileSystem=/ -pProtectSystem=full -pPrivateMounts=1 -pRootDirectory=/run/systemd-confinement/test -pRuntimeDirectory=systemd-confinement/test -pUMask=066 -pBindReadOnlyPaths=/nix/store -- $(realpath -e /bin/sh) -xc "PATH=$(dirname $(realpath -e $(which ls))); ls -la /; mkdir -v /test"
Running as unit: run-u2785.service

$ systemctl status run-u2785.service
× run-u2785.service - /nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash -xc PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin; ls -la /; mkdir -v /test
     Loaded: loaded (/run/systemd/transient/run-u2785.service; transient)
  Transient: yes
     Active: failed (Result: exit-code) since Wed 2024-05-01 21:41:44 CEST; 18s ago
   Duration: 10ms
    Process: 195345 ExecStart=/nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash -xc PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin; ls -la /; mkdir -v /test (code=exited, status=226/NAMESPACE)
   Main PID: 195345 (code=exited, status=226/NAMESPACE)
         IP: 0B in, 0B out
         IO: 0B read, 0B written
        CPU: 6ms

mai 01 21:41:44 oignon systemd[1]: Started /nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash -xc PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin; ls -la /; mkdir -v /test.
mai 01 21:41:44 oignon (bash)[195345]: run-u2785.service: Failed to set up mount namespacing: /run/systemd-confinement/test/etc: No such file or directory
mai 01 21:41:44 oignon (bash)[195345]: run-u2785.service: Failed at step NAMESPACE spawning /nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash: No such file or directory
mai 01 21:41:44 oignon systemd[1]: run-u2785.service: Main process exited, code=exited, status=226/NAMESPACE
mai 01 21:41:44 oignon systemd[1]: run-u2785.service: Failed with result 'exit-code'.

$ sudo systemd-run -P -pTemporaryFileSystem=/ -pProtectSystem=true -pPrivateMounts=1 -pRootDirectory=/run/systemd-confinement/test -pRuntimeDirectory=systemd-confinement/test -pUMask=066 -pBindReadOnlyPaths=/nix/store -- $(realpath -e /bin/sh) -xc "PATH=$(dirname $(realpath -e $(which ls))); ls -la /; mkdir -v /test"
Running as unit: run-u2786.service

$ systemctl status run-u2786.service
× run-u2786.service - /nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash -xc PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin; ls -la /; mkdir -v /test
     Loaded: loaded (/run/systemd/transient/run-u2786.service; transient)
  Transient: yes
     Active: failed (Result: exit-code) since Wed 2024-05-01 21:41:51 CEST; 48min ago
   Duration: 12ms
    Process: 195363 ExecStart=/nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash -xc PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin; ls -la /; mkdir -v /test (code=exited, status=226/NAMESPACE)
   Main PID: 195363 (code=exited, status=226/NAMESPACE)
         IP: 0B in, 0B out
         IO: 0B read, 0B written
        CPU: 10ms

mai 01 21:41:51 oignon systemd[1]: Started /nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash -xc PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin; ls -la /; mkdir -v /test.
mai 01 21:41:51 oignon (bash)[195363]: run-u2786.service: Failed to set up mount namespacing: /run/systemd-confinement/test/usr: No such file or directory
mai 01 21:41:51 oignon (bash)[195363]: run-u2786.service: Failed at step NAMESPACE spawning /nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash: No such file or directory
mai 01 21:41:51 oignon systemd[1]: run-u2786.service: Main process exited, code=exited, status=226/NAMESPACE
mai 01 21:41:51 oignon systemd[1]: run-u2786.service: Failed with result 'exit-code'.

mai 01 21:41:44 oignon systemd[1]: Started /nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash -xc PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin; ls -la /; mkdir -v /test.
mai 01 21:41:44 oignon (bash)[195345]: run-u2785.service: Failed to set up mount namespacing: /run/systemd-confinement/test/etc: No such file or directory
mai 01 21:41:44 oignon (bash)[195345]: run-u2785.service: Failed at step NAMESPACE spawning /nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash: No such file or directory
mai 01 21:41:44 oignon systemd[1]: run-u2785.service: Main process exited, code=exited, status=226/NAMESPACE
mai 01 21:41:44 oignon systemd[1]: run-u2785.service: Failed with result 'exit-code'
```

The goal is to deny write access to `/` to non-root users,
- `TemporaryFileSystem=/` gives us that through the ownership of `/` by `root` (instead of the service's user inherited from `RuntimeDirectory=`).
- `ProtectSystem=strict` gives us that by mounting `/` read-only (while keeping its ownership to the service's user).

To avoid the incompatibilities of `TemporaryFileSystem=/` mentionned above, I suggest to mount `/` read-only in all cases with `ReadOnlyPaths = [ "+/" ]`:

```console
$ sudo systemd-run -P -pReadOnlyPaths=+/ -pUser=999 -pPrivateMounts=1 -pRootDirectory=/run/systemd-confinement/test -pRuntimeDirectory=systemd-confinement/test -pUMask=066 -pBindReadOnlyPaths=/nix/store -- $(realpath -e /bin/sh) -xc "PATH=$(dirname $(realpath -e $(which ls))); ls -la /; mkdir -v /test"
Running as unit: run-u2780.service
+ PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin
+ ls -la /
total 0
drwxr-xr-x  11 999 999  220 May  1 19:40 .
drwxr-xr-x  11 999 999  220 May  1 19:40 ..
drwxr-xr-x  21   0   0 3800 May  1 16:52 dev
drwxr-xr-x   2   0   0   40 May  1 19:40 etc
drwxr-xr-x   3   0   0   60 May  1 19:40 nix
dr-xr-xr-x 321   0   0    0 May  1 19:40 proc
drwxr-x---   2   0   0   40 May  1 19:40 root
drwxrwxrwt   5   0   0  100 May  1 19:40 run
dr-xr-xr-x  13   0   0    0 Apr 29 18:38 sys
drwxr-xr-x   2   0   0   40 May  1 19:40 usr
drwxr-xr-x   2   0   0   40 May  1 19:40 var
+ mkdir -v /test
mkdir: cannot create directory '/test': Read-only file system
```
I *guess* this would require at least two changes to the current tests:
1. to no longer expect `root` to be able to write to some paths (like `/bin`) (at least not without first remounting `/` in read-write mode).
2. to no longer expect non-`root` users to fail to write to certain paths with a "permission denied" error code, but with a "read-only file system" error code.

Reply:
> it is however incompatible with both `ProtectSystem=full` or `ProtectSystem=true`:

ah, good to know, will add parametrised tests for those options as well.

> To avoid the incompatibilities of `TemporaryFileSystem=/` mentionned above, I suggest to mount `/` read-only in all cases with `ReadOnlyPaths = [ "+/" ]`:

:+1: - will implement that later today or tomorrow.